Added DeleteBlob operations with integration and unit tests

BrianMcBrayer opened the pull request:
* Pressed Ctrl+K, Ctrl+D (oops, lots of whitspace gone... makes this PR look bad)
* Added DeleteBlob and DeleteBlobAsync
* Added Delete unit tests
* Added Delete integration tests

tarwn commented:
DeleteBlobResponse: You appear to have some unnecessary copy+paste leftovers. I don't see Metadata or body content mentioned in the API page for DeleteBlob, so unless I misread the ogic to handle those aren't necessary for DeleteBlob

Unit Tests: There are some conventions in place for the unit tests that I suppose I should document clearly somewhere.
1. All (most) of the tests are written using a 3-part naming pattern: FunctionUnderTest_Scenario_Expectation
This communicates exactly what is broken when a unit tests fails and you just see it in a build list and makes it easy to scan through the list of collapsed tests to find one that you remember the intent for but not the details. It also forces you to define your intent clearly, which may result in you changing what you test.

2. AAA Pattern
All of the tests are written using the AAA pattern and I attempted to space them all out to make it clear what belonged to the seperate Arrange, Act, and Assert stages. When a test was too long or needed extra whitespace tat would not make this obvious, I added comments above each section. When you're coming in to update a test, you're often looking for the Arrange section nd it's easy to find. When you're troubleshooting a broken test, you often start at Assert, then look at Act, then look at Arrange.

3. BlobServiceClientTests - DeleteBlob_ExistingBlob and DeleteBlobAsync_ExistingBlob
* Naming for the two tests per 1
* Format for the tests per 2
* The pre-exists check is basically testing our setup code and/or the MS Azure SDK and the already tested GetBlob method, so I would remove it (they can test that on their own and we have plenty on GteBlob already). If this shared logic failed, we have numerous other tests that would fail and paint a pretty clear picture of what was wrong. If we wanted to keep this assertion, we would split this into two tests, one having the intent to test the setup code and one having the intent to test the DeleteBlobRequest
* In tests, try to only use the BasicAzureStorageSDK functions as the function under test and do any setup or assertions using the MS Azure SDK. This reduces the risk of introducing an error (likely through copy and paste in multiple modules) that all of the functions do equally wrong but appear to be doing right in the test. It also means that if we break the GetBlob function, only the GetBlob tests will fail (and the DeleteBlob will continue working). This means it is clear when tests fail for multiple functions under tests that we have done something lower-level that affects multiple operatoins, versus all the tests failing for a single function (like GetBlob) which indicates it is likely somethign specific to just that one function.
*  In the other tests, I refactored down common checks like AssertBlobExists into a function so the intent would be clear in the tests and the implementation would be common across any other tests that needed it. When you glance at the tail end of one of those tests, it's a single line and clear what the intent is. You've copied the same 10 lines of logic to verify the blob has been deleted into both tests. It would be cleaner to move this to a separate function. There is an AssertContainerExists, AssertContainerDoesNotExist, and AssertBlobExists, the pattern should be pretty clear to create a AssertBlobDoesNotExist

BrianMcBrayer commented:
Oh wow. Lots to do. I'll have it cleaned up in a jiffy.

Sent from my Windows Phone
________________________________
From: Eli Weinstock-Herman<mailto:notifications@github.com>
Sent: ‎8/‎17/‎2015 10:19 AM
To: tarwn/BasicAzureStorageSDK<mailto:BasicAzureStorageSDK@noreply.github.com>
Cc: Brian McBrayer<mailto:brandmcbrayer@live.com>
Subject: Re: [BasicAzureStorageSDK] Added DeleteBlob operations with integration and unit tests (#5)

DeleteBlobResponse: You appear to have some unnecessary copy+paste leftovers. I don't see Metadata or body content mentioned in the API page for DeleteBlob, so unless I misread the ogic to handle those aren't necessary for DeleteBlob

Unit Tests: There are some conventions in place for the unit tests that I suppose I should document clearly somewhere.
1. All (most) of the tests are written using a 3-part naming pattern: FunctionUnderTest_Scenario_Expectation
This communicates exactly what is broken when a unit tests fails and you just see it in a build list and makes it easy to scan through the list of collapsed tests to find one that you remember the intent for but not the details. It also forces you to define your intent clearly, which may result in you changing what you test.

2. AAA Pattern
All of the tests are written using the AAA pattern and I attempted to space them all out to make it clear what belonged to the seperate Arrange, Act, and Assert stages. When a test was too long or needed extra whitespace tat would not make this obvious, I added comments above each section. When you're coming in to update a test, you're often looking for the Arrange section nd it's easy to find. When you're troubleshooting a broken test, you often start at Assert, then look at Act, then look at Arrange.

3. BlobServiceClientTests - DeleteBlob_ExistingBlob and DeleteBlobAsync_ExistingBlob
* Naming for the two tests per 1
* Format for the tests per 2
* The pre-exists check is basically testing our setup code and/or the MS Azure SDK and the already tested GetBlob method, so I would remove it (they can test that on their own and we have plenty on GteBlob already). If this shared logic failed, we have numerous other tests that would fail and paint a pretty clear picture of what was wrong. If we wanted to keep this assertion, we would split this into two tests, one having the intent to test the setup code and one having the intent to test the DeleteBlobRequest
* In tests, try to only use the BasicAzureStorageSDK functions as the function under test and do any setup or assertions using the MS Azure SDK. This reduces the risk of introducing an error (likely through copy and paste in multiple modules) that all of the functions do equally wrong but appear to be doing right in the test. It also means that if we break the GetBlob function, only the GetBlob tests will fail (and the DeleteBlob will continue working). This means it is clear when tests fail for multiple functions under tests that we have done something lower-level that affects multiple operatoins, versus all the tests failing for a single function (like GetBlob) which indicates it is likely somethign specific to just that one function.
*  In the other tests, I refactored down common checks like AssertBlobExists into a function so the intent would be clear in the tests and the implementation would be common across any other tests that needed it. When you glance at the tail end of one of those tests, it's a single line and clear what the intent is. You've copied the same 10 lines of logic to verify the blob has been deleted into both tests. It would be cleaner to move this to a separate function. There is an AssertContainerExists, AssertContainerDoesNotExist, and AssertBlobExists, the pattern should be pretty clear to create a AssertBlobDoesNotExist

---
Reply to this email directly or view it on GitHub:
https://github.com/tarwn/BasicAzureStorageSDK/pull/5#issuecomment-131834756

BrianMcBrayer commented:
Updated per comments
